### PR TITLE
polycalc: new port in math

### PR DESCRIPTION
**Pull request by barracuda156**
#### Description

New port: https://github.com/Zeta611/polycalc

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.6.8 Server
Xcode 3.2.6

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

File: math/polycalc/Portfile [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+                    port:bison \
+                    port:flex
+
+compiler.c_standard 2017
+
+post-patch {
+    reinplace "s,gcc,${configure.cc}," ${worksrcpath}/Makefile
+}
+
+build.cmd           make
```
Comment (harens) [inline math/polycalc/Portfile:30]:
```suggestion
```

I think `build.cmd` should already be set as make.

File: math/polycalc/Portfile [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+depends_build-append \
+                    port:bison \
+                    port:flex
+
+compiler.c_standard 2017
+
+post-patch {
+    reinplace "s,gcc,${configure.cc}," ${worksrcpath}/Makefile
+}
```
Comment (harens) [inline math/polycalc/Portfile:28]:
```suggestion
```

Generally, this sort of thing is better as a patch file. However, I don't think it's needed here. The makefile portgroup should add `CC="${configure.cc}"` as a build arg.

File: math/polycalc/Portfile (hunk 1 above)
Comment (barracuda156) [reply math/polycalc/Portfile:30]: Did not work for me initially if removed, but let me try again (there was a fix with upstream for our clang builds, I am not sure if I tried to remove this line after that).

Comment (barracuda156) [reply math/polycalc/Portfile:30]: @harens No, it fails without this. I dunno why :) But I normally do not add something that should be automatic without a reason (_well, I try to, at least_).
Here is what happens with both build settings removed:
```
--->  Configuring polycalc
--->  Building polycalc
Executing:  cd "/opt/local/var/macports/build/_opt_PPCRosettaPorts_math_polycalc/polycalc/work/polycalc-8c6922fab48f8e23ac4f02d00cd27cb271a02965" && /usr/bin/make -j8 -w all CC="/opt/local/var/macports/build/_opt_PPCRosettaPorts_math_polycalc/polycalc/work/compwrap/cc/opt/local/bin/gcc-mp-12" CXX="/opt/local/var/macports/build/_opt_PPCRosettaPorts_math_polycalc/polycalc/work/compwrap/cxx/opt/local/bin/g++-mp-12" OBJC="/opt/local/var/macports/build/_opt_PPCRosettaPorts_math_polycalc/polycalc/work/compwrap/objc/opt/local/bin/gcc-mp-12" OBJCXX="/opt/local/var/macports/build/_opt_PPCRosettaPorts_math_polycalc/polycalc/work/compwrap/objcxx/opt/local/bin/g++-mp-12" FC="/opt/local/var/macports/build/_opt_PPCRosettaPorts_math_polycalc/polycalc/work/compwrap/fc/opt/local/bin/gfortran-mp-12" F77="/opt/local/var/macports/build/_opt_PPCRosettaPorts_math_polycalc/polycalc/work/compwrap/f77/opt/local/bin/gfortran-mp-12" F90="/opt/local/var/macports/build/_opt_PPCRosettaPorts_math_polycalc/polycalc/work/compwrap/f90/opt/local/bin/gfortran-mp-12" INSTALL="/usr/bin/install -c" 
make: Entering directory `/opt/local/var/macports/build/_opt_PPCRosettaPorts_math_polycalc/polycalc/work/Zeta611-polycalc-8c6922f'
make: *** No rule to make target `all'.  Stop.
```
And this with build.cmd removed and build.target retained:
```
./build/src/poly.tab.c:372:54: error: token "\" is not valid in preprocessor expressions
  372 | #  if (defined __cplusplus && ! defined EXIT_SUCCESS \
      |                                                      ^
./build/src/poly.tab.c:372: error: unterminated #if
  372 | #  if (defined __cplusplus && ! defined EXIT_SUCCESS \
      | 
./build/src/poly.tab.c:356: error: unterminated #else
  356 | # ifdef YYSTACK_ALLOC
      | 
./build/src/poly.tab.c:328: error: unterminated #if
  328 | #if !defined yyoverflow
      | 
make: *** [build/./src/poly.y.o] Error 1
make: *** Waiting for unfinished jobs....
```
With both in place, it builds fine.

File: math/polycalc/Portfile (hunk 2 above)
Comment (barracuda156) [reply math/polycalc/Portfile:28]: You are right, this appears unneeded. I have removed it, and still correct GCC is chosen. I will delete this block from the portfile now.

Comment (barracuda156) [reply math/polycalc/Portfile]: Worked for me, but fails on CI:
```
/opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_math_polycalc/polycalc/work/compwrap/cc/usr/bin/clang -I./src -MMD -MP -O3 -Wall -Wextra -Wpedantic -std=c17 -Wno-implicit-function-declaration -c build/./src/poly.tab.c -o build/./src/poly.y.o
[362](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:365)
  src/poly.y:180:1: error: unknown type name 'near'
[363](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:366)
  near line %d\n", progname, msg, lineno);
[364](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:367)
  ^
[365](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:368)
  src/poly.y:180:10: error: expected ';' after top level declarator
[366](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:369)
  near line %d\n", progname, msg, lineno);
[367](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:370)
           ^
[368](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:371)
           ;
[369](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:372)
  src/poly.y:180:15: warning: missing terminating '"' character [-Winvalid-pp-token]
[370](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:373)
  near line %d\n", progname, msg, lineno);
[371](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:374)
                ^
[372](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:375)
  src/poly.y:182:1: error: extraneous closing brace ('}')
[373](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:376)
  }
[374](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:377)
  ^
[375](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:378)
  1 warning and 3 errors generated.
[376](https://github.com/macports/macports-ports/actions/runs/3838206817/jobs/6534389403#step:8:379)
  make: *** [build/./src/poly.y.o] Error 1
```

**Comment by barracuda156** [comment]
@harens I have added a patch, so that `gcc` is not reinplaced directly, but it seems all those manual settings are needed, at least on some systems (well, both locally for me and on CI, at least).

**Comment by barracuda156** [comment]
UPD. This is a mystery: now macos-11 CI fail, but literally nothing changed in the portfile beside the patch – which must be strictly inconsequential:
```
src/poly.y:180:10: error: expected ';' after top level declarator
  near line %d\n", progname, msg, lineno);
           ^
           ;
  src/poly.y:180:15: warning: missing terminating '"' character [-Winvalid-pp-token]
  near line %d\n", progname, msg, lineno);
                ^
  src/poly.y:182:1: error: extraneous closing brace ('}')
  }
  ^
  1 warning and 3 errors generated.
```

**Comment by barracuda156** [comment]
Well, CI are broken now by this commit, apparently: https://github.com/macports/macports-ports/commit/eb481fbca2d51b4f9366d397dd2bad2971fdbdf7

```
Creating port index in /Users/runner/work/macports-ports/macports-ports/ports
  Failed to parse file math/octave-statistics-bootstrap/Portfile: can't set "depends_lib": invalid depspec: octave-statistics
```

File: math/polycalc/Portfile (hunk 1 above)
Comment (harens) [reply math/polycalc/Portfile]: Fair enough, thanks for checking :)

Comment (barracuda156) [reply math/polycalc/Portfile]: @harens I honestly stopped understanding this build. I tried today to rebuild it few times, and it did work without patch and most of manual settings. No idea what changed. For now, I removed extra settings, let us see if it builds on bots.
(Now CI are broken due to unrelated issue.)

Comment (harens) [reply math/polycalc/Portfile]: It's all been very unusual. I have no idea why it sometimes builds and then fails when nothing's changed. Thanks for persisting though.

**Comment by barracuda156** [comment]
@harens Okay, so now with minimal portfile (patch removed, `build.cmd` removed`) we get this on CI:
```
./build/src/poly.tab.c:815:1: error: unterminated /* comment
  /* YYMAXDEPTH -- maximum size the stacks can grow to (effective only
  ^
  ./build/src/poly.tab.c:816:33: warning: no newline at end of file [-Wnewline-eof]
     if the built-in stack extensi
                                  ^
  1 warning and 1 error generated.
  make: *** [build/./src/poly.y.o] Error 1
```

**Comment by barracuda156** [comment]
@harens And we are back to one CI pass, one CI fails, _but in reverse:_ https://github.com/macports/macports-ports/pull/17179#issuecomment-1370955113

Locally I also see no consistency: the build randomly succeeds or fails with strictly identical settings.

Let me just close this PR.

**Comment by harens** [comment]
> _but in reverse_

Because why not `¯\_(ツ)_/¯`. This was one weird port.

> Let me just close this PR.

Thanks anyway though @barracuda156 👍

**Comment by kencu** [comment]
a common cause for such inconsistent build issues is parallel building

**Comment by barracuda156** [comment]
@kencu Thank you! You saved the port :)

P. S. This reminded me a nightmare with trying to fix `nhc98` for months, and eventually I realized I just needed `use_parallel_build no` for everything to work.

**Comment by barracuda156** [comment]
> > _but in reverse_
> 
> Because why not `¯\_(ツ)_/¯`. This was one weird port.

@harens So you were right, no need in manual patching, what we missed was parallel_build.

**Comment by harens** [comment]
> @harens So you were right, no need in manual patching, what we missed was parallel_build.

That's great news! Thanks @kencu for the tip.
